### PR TITLE
Add urbandictionary command

### PR DESCRIPTION
**Pull request by Miroito**

### CODE REVIEW

File: src/modules/commands/discord/urbandictionary.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+export default new Command<discord.Message>({
+  keyword: 'urbandictionary',
+  description: 'get the modern definition of a word or expression',
+  help: 'Usage: `{prefix}urbandictionary` - get the modern definition of a word or expression',
+  callback: async ({ args, reply, env }) => {
+    if (args.length === 0) return reply(env.command.getHelp(env.handler))
+
+    const word = args.join(' ') //Merge all the arguments for expressions
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:28]: Getting one argument with multiple words is built into CommandHandler with the `lastArgumentNumber` option. Check how it's used in the math command.

File: src/modules/commands/discord/urbandictionary.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    const term = word.toLowerCase()
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:30]: UrbanDictionary API is case sensitive, you shouldn't convert the term to lowercase.

File: src/modules/commands/discord/urbandictionary.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      .json()
+      .then((data) => {
+        if (!isRecord(data)) {
+          return Promise.reject(new Error(`Unexpected query return type`))
+        }
+        if (data.error) {
+          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+          return Promise.reject(new Error(`Query error: ${data.error}`))
+        }
+        if (data.errorCode) {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:43]: `data.error` and `data.errorCode` are error constructs from the Daybreak API that don't apply here. Either check what (if any) errors UD returns or just delete it.

File: src/modules/commands/discord/urbandictionary.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          return Promise.reject(
+            new Error(`${collectionListName} is not an array`),
+          )
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+        return camelcaseKeys(definitions, { deep: true })
+      })) as definition[]
+
+    // eslint-disable-next-line no-useless-escape
+    const re = /\[(\w+(\s\w+)*?)\]/g
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:64]: I don't see any unnecessary escape here? Try removing the rule if it's even needed.

File: src/modules/commands/discord/urbandictionary.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    if (list.length === 0)
+      if (args.length > 1)
+        return reply(`There is no definition for the expression **${word}**.`)
+      else return reply(`There is no definition for the word **${word}**.`)
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:69]: The four lines above are quite difficult to read without curly braces. Please add them.

File: src/modules/commands/discord/urbandictionary.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import { Command } from '@commands/CommandHandler'
+import got from 'got'
+import { isRecord } from '@app/validators/object'
+import camelcaseKeys from 'camelcase-keys'
+import discord from 'discord.js'
+
+type definition = {
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:7]: Types are conventionally written in PascalCase, i.e. please change to Definition.

File: src/modules/commands/discord/urbandictionary.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  author: string
+  word: string
+  defid: number
+  currentVote: string
+  writtenOn: string
+  example: string
+  thumbsDown: number
+}
+
+export default new Command<discord.Message>({
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:21]: `commands/discord` are meant for commands that use Discord specific features and need the `raw` parameter set to `discord.Message`. However you do not use `raw` here, so you should move it to `commands/common`.

File: src/modules/commands/discord/urbandictionary.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        if (data.error) {
+          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+          return Promise.reject(new Error(`Query error: ${data.error}`))
+        }
+        if (data.errorCode) {
+          // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+          return Promise.reject(new Error(`Query error: ${data.errorCode}`))
+        }
+        const collectionListName = 'list'
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:47]: Unnecessary intermediate constant for `'list'`, simply access `data.list`.

File: src/modules/commands/discord/urbandictionary.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    if (list.length === 0)
+      if (args.length > 1)
+        return reply(`There is no definition for the expression **${word}**.`)
+      else return reply(`There is no definition for the word **${word}**.`)
+    list = list.sort((a: definition, b: definition) => {
+      if (b.thumbsUp > a.thumbsUp) return 1
+      else return -1
+    })
+    return reply(
+      `**Word:** ${
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:75]: I like how above you use either "word" or "expression" depending on the number of words. You probably want that here as well.

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/modules/commands/discord/urbandictionary.ts (hunk 4 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:64]: Ye oups something from testing

File: src/modules/commands/discord/urbandictionary.ts (hunk 7 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:21]: I want to make it into an embed later, should I still move it?

Comment (vostrnad) [reply src/modules/commands/discord/urbandictionary.ts:21]: All good then.

File: src/modules/commands/discord/urbandictionary.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  help: 'Usage: `{prefix}urbandictionary` - get the modern definition of a word or expression',
+  options: {
+    lastArgNumber: 1,
+  },
+  callback: async ({ args, reply, env }) => {
+    if (args.length === 0) return reply(env.command.getHelp(env.handler))
+
+    const word = args[0]
+
+    const term = word
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:33]: Why do you have `word` and `term` when they are the same? (I prefer term)

File: src/modules/commands/discord/urbandictionary.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  currentVote: string
+  writtenOn: string
+  example: string
+  thumbsDown: number
+}
+
+export default new Command<discord.Message>({
+  keyword: 'urbandictionary',
+  description: 'get the modern definition of a word or expression',
+  help: 'Usage: `{prefix}urbandictionary` - get the modern definition of a word or expression',
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:24]: Help should include argument description i.e. `<expression>`

File: src/modules/commands/discord/urbandictionary.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+        return camelcaseKeys(definitions, { deep: true })
+      })) as Definition[]
+
+    const re = /\[(\w+(\s\w+)*?)\]/g
+
+    let wordString: string
+    if (word.includes(' ')) wordString = 'expression'
+    else wordString = 'word'
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:58]: There is much better syntax for this:
```ts
const wordString = word.includes(' ') ? 'expression' : 'word'
```

File: src/modules/commands/discord/urbandictionary.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    if (list.length === 0) {
+      if (args.length > 1)
```
Comment (vostrnad) [inline src/modules/commands/discord/urbandictionary.ts:61]: This branching isn't needed anymore (you have wordString).

**Review by vostrnad — CHANGES_REQUESTED** [review]

File: src/modules/commands/discord/urbandictionary.ts (hunk 10 above)
Comment (Miroito) [reply src/modules/commands/discord/urbandictionary.ts:33]: hahahahha

**Review by vostrnad — APPROVED** [review]

**Pull request merged**